### PR TITLE
Make js version usable in npm

### PR DESCRIPTION
**Pull request by michal-kurz**

### CODE REVIEW

File: javascript/diff_match_patch_uncompressed.js [hunk 1]
```diff
@@ -2407,18 +2418,8 @@ diff_match_patch.patch_obj.prototype.toString = function() {
...
-this['DIFF_DELETE'] = DIFF_DELETE;
-/** @suppress {globalThis} */
-this['DIFF_INSERT'] = DIFF_INSERT;
-/** @suppress {globalThis} */
-this['DIFF_EQUAL'] = DIFF_EQUAL;
+module.exports = diff_match_patch;
+module.exports['diff_match_patch'] = diff_match_patch;
+module.exports['DIFF_DELETE'] = DIFF_DELETE;
+module.exports['DIFF_INSERT'] = DIFF_INSERT;
+module.exports['DIFF_EQUAL'] = DIFF_EQUAL;
```
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:2414]: was this change intentional?

Comment (michal-kurz) [reply javascript/diff_match_patch_uncompressed.js:2414]: This has nothing to do with the fix - it's an older change I made when prototyping a while I started using my fork as a npm package. I will clean this up :)

File: javascript/diff_match_patch_uncompressed.js [hunk 2]
```diff
@@ -1776,12 +1776,23 @@ diff_match_patch.prototype.patch_addContext_ = function(patch, text) {
 
   // Add the prefix.
   var prefix = text.substring(patch.start2 - padding, patch.start2);
+  
+  if(this.isLowSurrogate(prefix[0])) {
+    prefix = text.substring(patch.start2 - padding - 1 , patch.start2);
+  }
```
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js]: just a note here: we can avoid a double string copy if we check for the surrogate half before substringing (same goes for the suffix).

```js
// Avoid splitting on non-character boundaries when pulling in context.
var prefix = this.isLowSurrogate(text[patch.start2 - padding])
	? text.substring(patch.start2 - padding - 1, patch.start2)
	: text.substring(patch.start2 - padding, patch.start2);
```

File: javascript/diff_match_patch_uncompressed.js (hunk 1 above)
Comment (michal-kurz) [reply javascript/diff_match_patch_uncompressed.js:2414]: same goes for the package.json

**Comment by dmsnell** [comment]
@michal-kurz good catch on this. there's something else to consider here: considering that we're pulling in extra characters into the patch context, I guess there's still a possibility we might crash if the input document contains a series of surrogate halves. it might be that the existing fixup makes the same assumption, that if we shift one character we'll not find another lone low or high surrogate. just wanted to point it out.

File: javascript/diff_match_patch_uncompressed.js (hunk 2 above)
Comment (michal-kurz) [reply javascript/diff_match_patch_uncompressed.js]: Good point, thank you, I will change this tomorrow, along with the cleanup i mentioned before. I'm off to bed for now :)

Comment (michal-kurz) [reply javascript/diff_match_patch_uncompressed.js]: @dmsnell This branch (and PR) was not a correct place to put my fix - I moved it over here: https://github.com/feedyou-ai/diff-match-patch/pull/4

File: javascript/diff_match_patch_uncompressed.js (hunk 1 above)
Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:2414]: I like the change in package.json, but I'm nervous about these lines and if they are okay. did you verify that everything comes through Closure Compiler as it did before?

I guess one simple test is simply to compile with and without this change since no other code is different. If the output compressed bundle is identical then we should be fine.

File: javascript/diff_match_patch_uncompressed.js (hunk 2 above)
Comment (michal-kurz) [reply javascript/diff_match_patch_uncompressed.js]: @dmsnell Just a quick followup: I had to also do this: https://github.com/feedyou-ai/diff-match-patch/pull/7

because `text[patch.start2 - padding]` can be out of bounds of `text` and return undefined :)
